### PR TITLE
Write replay measurements into a JSON file

### PR DESCRIPTION
**Pull request by marius-pelegrin-arm**
Results from `--measurement-frame-range` can now be written to a JSON file using the option `--measurement-file`.
In addition, a manager has been created to easily write to the JSON structure from anywhere in the code.

This option is necessary to retrieve measurements in android environments where logcat is not available. It also makes parsing of the measurements by a script easier and allow more detailed measurements without flooding the logcat.

Example of JSON output:

```json
{
    "application": {
        "capture_name": "/sdcard/capture.gfxr"
    },
    "frame_ranges": [
        {
            "duration": 2.365642154,
            "end_frame": 200,
            "end_time_monotonic": 15666.74495701,
            "fps": 42.27182028816688,
            "frame_count": 100,
            "frame_durations": [
                17316846,
                26862231,
                24499077,
                24042769,

                ...

                19480077,
                16187384,
                16081154,
                18678846,
                18439692
            ],
            "start_frame": 100,
            "start_time_monotonic": 15664.379314856
        }
    ]
}
```

### CODE REVIEW

File: tools/tool_settings.h [hunk 1]
```diff
@@ -39,6 +39,7 @@
 #include "generated/generated_vulkan_decoder.h"
 #include "util/argument_parser.h"
 #include "util/logging.h"
+#include "util/measurement_manager.h"
```
Comment (panos-lunarg) [inline tools/tool_settings.h:42]: I think this needs to be included in the *_main.cpp files instead of here.

Comment (marius-pelegrin-arm) [reply tools/tool_settings.h]: Done !

**Comment by andrew-lunarg** [comment]
Looks like you don't include [\<optional\>](https://en.cppreference.com/w/cpp/header/optional)

**Comment by andrew-lunarg** [comment]
> CI gfxreconstruct build # 2916 failed.

Failures like this are down to a failing machine in CI being discussed internally. Probably not going to be resolved today, FYI.

**Comment by lunarpapillo** [comment]
The failure in #2916 was due to a NAS mount failure.  I've requeued.

**Review by andrew-lunarg — CHANGES_REQUESTED** [review]
The pattern I see reading through the code is that the main functions of the replay app need to init the JSON tree used to store measurements and to communicate the name of the measurement file to the fps_info.cpp file from which it is written. They do so through a single global variable of type `MeasurementFile` which bundles the filename and tree root. Adding a whole new compilation unit and two classes to get this information to where it is used seems over the top since the filename could be passed to the `FpsInfo` constructor and the tree could be created late and locally just before outputting it without going through global scope.

# Observations

## MeasurementManager
`MeasurementManager` is a pair of static data items and some static functions. It is mostly equivalent to a namespace. It is also an obfuscation of using a global variable to move data between points in the program. I don’t think it really does much “managing”.

## MeasurementFile
`MeasurementFile` sounds like a file abstraction but it is not a file abstraction. A file is created and destroyed only within the local scope of its destructor. It bundles a filename and a JSON tree so they can be moved from where they are determined and initialised to where they are used.
Note, `MeasurementFile::~MeasurementFile()` performs potentially throwing operations such as allocating a large std::string and more likely to fail, performing filesystem file deletion. Destructors shouldn’t ever throw (https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rc-dtor-fail).

# Suggested Design Alternatives

## Minimum Effort to Implement Feature In PR

* `FpsInfo` gets a member string filename naming the file it will later save measurements into.
* `void FpsInfo::EndFrame` builds the JSON tree in a local variable with the file ops all also locally scoped and writes it to a file using the filename it got at construction time.
* No class MeasurementFile
* No class MeasurementManager

The only reason to try to be more elaborate would be to lay the groundwork for a follow-up PR that needed a persistent JSON tree of measurements because the length of `"frame_ranges"` could be greater than 1 (which the flags defined by this PR and code implementing it do not allow). However, a future PR expanding the feature would need to touch lots of this code anyway so we may as well make the current PR as simple as possible.

## More Effort Assuming "frame_ranges":[n]

* `FpsInfo` gets a member string filename naming the file it will later save measurements into.
* `FpsInfo` gets a json tree member to buffer in-memory measurements into.
* `void FpsInfo::EndFrame` appends to the member JSON tree.
* `void FpsInfo::EndFrame` dumps the JSON tree to a file which it creates there and then when the last frame range is detected (for this PR the first range is also the last range).
* No class MeasurementFile
* No class MeasurementManager

It would be pointless to do this without any  way to measure multiple ranges. I say, do the previous suggestion.


# Conclusions / Actions

We would like to bring in the feature but with the implementation tweaked to be more simple, as per my first suggested design above. We would be happy to do the work to make those changes and raise a PR against https://github.com/marius-pelegrin-arm/GFXReconstruct/tree/measurements-file. Would either you or us making this change work for ARM?

Additionally, would changing the JSON field `"frame_ranges"` (an array of objects holding measurements that is always length one) to `"frame_range"` (a single object with the measurements in it) work for you?

File: framework/util/measurement_manager.h [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+
+  private:
+    std::string    file_name_;
+    nlohmann::json root_;
+};
+
+class MeasurementManager
+{
+  public:
+    static void Open(const std::string& file_name);
```
[the remark this answers is not in the document]
Comment (marius-pelegrin-arm) [reply framework/util/measurement_manager.h]: Removed

File: framework/util/measurement_manager.cpp [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    std::string json_string = root_.dump(4); // TODO: '--measurement-file-pretty' option with/without identation ?
+
+    size_t size_written = platform::FileWrite(json_string.data(), 1, json_string.size(), file_pointer);
+
+    platform::FileClose(file_pointer);
+
+    // It either writes a fully valid file, or it doesn't write anything !
+    if (size_written != json_string.size())
+    {
+        std::remove(file_name_.c_str());
```
[the remark this answers is not in the document]
Comment (marius-pelegrin-arm) [reply framework/util/measurement_manager.cpp]: Moved to a normal function

File: framework/util/measurement_manager.cpp [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+    FILE*   file_pointer = nullptr;
+    int32_t result       = platform::FileOpen(&file_pointer, file_name_.c_str(), "w");
+    if (result != 0)
+    {
+        GFXRECON_LOG_ERROR("Failed to open measurements file '%s' (Error %i).", file_name_.c_str(), result);
+        GFXRECON_LOG_ERROR("%s", std::strerror(result));
+        return;
+    }
+
+    std::string json_string = root_.dump(4); // TODO: '--measurement-file-pretty' option with/without identation ?
```
[the remark this answers is not in the document]
Comment (marius-pelegrin-arm) [reply framework/util/measurement_manager.cpp]: Moved to a normal function

File: framework/util/measurement_manager.cpp [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+
+MeasurementFile::MeasurementFile(const std::string& file_name) :
+    file_name_(file_name),
+    root_({ { "application", nlohmann::json::object() }, { "frame_ranges", nlohmann::json::array() } })
+{
+    // To avoid thinking an ancient file is the result of this run
+    std::remove(file_name.c_str());
+}
+
+MeasurementFile::~MeasurementFile()
```
[the remark this answers is not in the document]
Comment (marius-pelegrin-arm) [reply framework/util/measurement_manager.cpp]: Done in FpsInfo::EndFrame

File: tools/replay/android_main.cpp [hunk 6]
```diff
@@ -114,6 +115,14 @@ void android_main(struct android_app* app)
                 uint32_t                               start_frame, end_frame;
                 bool has_mfr = GetMeasurementFrameRange(arg_parser, start_frame, end_frame);
 
+                if (has_mfr)
+                {
+                    std::string measurement_file_name;
+                    GetMeasurementFilename(arg_parser, measurement_file_name);
+                    gfxrecon::util::MeasurementManager::Open(measurement_file_name);
```
[the remark this answers is not in the document]
Comment (marius-pelegrin-arm) [reply tools/replay/android_main.cpp]: Removed

File: framework/util/measurement_manager.cpp [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+#include "util/logging.h"
+
+GFXRECON_BEGIN_NAMESPACE(gfxrecon)
+GFXRECON_BEGIN_NAMESPACE(util)
+
+std::optional<MeasurementFile> MeasurementManager::file_;
+
+MeasurementFile::MeasurementFile(const std::string& file_name) :
+    file_name_(file_name),
+    root_({ { "application", nlohmann::json::object() }, { "frame_ranges", nlohmann::json::array() } })
```
[the remark this answers is not in the document]
Comment (marius-pelegrin-arm) [reply framework/util/measurement_manager.cpp]: Changed to "frame_range" which is a simple object and not an array of objects

File: framework/util/measurement_manager.cpp [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+{
+    // To avoid thinking an ancient file is the result of this run
+    std::remove(file_name.c_str());
```
[the remark this answers is not in the document]
Comment (marius-pelegrin-arm) [reply framework/util/measurement_manager.cpp]: Yes we do: https://en.cppreference.com/w/cpp/io/c/remove

There is no overload of this function using std::string or std::filesystem::path

File: framework/graphics/fps_info.cpp [hunk 9]
```diff
@@ -90,19 +92,40 @@ void FpsInfo::BeginFrame(uint64_t frame)
...
+            util::MeasurementManager::WriteFrameRange("start_frame", measurement_start_frame_);
+            util::MeasurementManager::WriteFrameRange("end_frame", measurement_end_frame_);
+            util::MeasurementManager::WriteFrameRange("frame_count", total_frames);
+            util::MeasurementManager::WriteFrameRange("start_time_monotonic", start_time);
+            util::MeasurementManager::WriteFrameRange("end_time_monotonic", end_time);
+            util::MeasurementManager::WriteFrameRange("duration", diff_time);
+            util::MeasurementManager::WriteFrameRange("fps", fps);
+            util::MeasurementManager::WriteFrameRange("frame_durations", frame_durations_);
+
+            util::MeasurementManager::SaveAndClose();
```
[the remark this answers is not in the document]
Comment (marius-pelegrin-arm) [reply framework/graphics/fps_info.cpp]: Done

**Comment by marius-pelegrin-arm** [comment]
Everything should be resolved. The complexity was introduced because "what if": What if we have multiple measurement ranges, what if we want to write more measurements from multiple places of the code and not just from FpsInfo etc.

Having something simple as a first step and making it more complicated if we have the need is completely understandable. This commit should be simpler.

File: framework/graphics/fps_info.cpp [hunk 10]
```diff
@@ -54,16 +56,26 @@ WriteFpsToConsole(const char* prefix, uint64_t start_frame, uint64_t end_frame,
...
-                 uint64_t measurement_end_frame,
-                 bool     has_measurement_range,
-                 bool     quit_after_range,
-                 bool     flush_measurement_range) :
+FpsInfo::FpsInfo(uint64_t           measurement_start_frame,
+                 uint64_t           measurement_end_frame,
+                 bool               has_measurement_range,
+                 bool               quit_after_range,
+                 bool               flush_measurement_range,
+                 const std::string& measurement_file_name) :
```
Comment (andrew-lunarg) [inline framework/graphics/fps_info.cpp:64]: You could make this a `const string_view` for flexibility if you wanted but this is not a requirement.

File: framework/graphics/fps_info.cpp [hunk 11]
```diff
@@ -90,19 +102,71 @@ void FpsInfo::BeginFrame(uint64_t frame)
...
+            // Save measurements to file
+            if (!measurement_file_name_.empty())
+            {
+                double   start_time   = util::datetime::ConvertTimestampToSeconds(measurement_start_time_);
+                double   end_time     = util::datetime::ConvertTimestampToSeconds(measurement_end_time_);
+                double   diff_time    = GetElapsedSeconds(measurement_start_time_, measurement_end_time_);
+                uint64_t total_frames = measurement_end_frame_ - measurement_start_frame_;
+                double   fps          = static_cast<double>(total_frames) / diff_time;
+
+                nlohmann::json file_content = { { "frame_range",
```
Comment (andrew-lunarg) [inline framework/graphics/fps_info.cpp:133]: Your previous version had an object, something like `"application":{ "capture_name": "a/file/name"}` in here too. Just checking the deletion is deliberate (it was added to the tree in the manager previously).

File: framework/graphics/fps_info.cpp [hunk 12]
```diff
@@ -90,19 +102,71 @@ void FpsInfo::BeginFrame(uint64_t frame)
...
+                                                    { "end_time_monotonic", end_time },
+                                                    { "duration", diff_time },
+                                                    { "fps", fps },
+                                                    { "frame_durations", frame_durations_ } } } };
+
+                FILE*   file_pointer = nullptr;
+                int32_t result       = util::platform::FileOpen(&file_pointer, measurement_file_name_.c_str(), "w");
+                if (result == 0)
+                {
+                    const std::string json_string = file_content.dump(4);
```
Comment (andrew-lunarg) [inline framework/graphics/fps_info.cpp:147]: We have a `kJsonIndentWidth` if you want to be consistent with json output elsewhere. Up to you.

**Review by andrew-lunarg — APPROVED** [review]
Thank you for your patient efforts in getting this in shape. It looks good to me now. There are a couple of comments for you to take or leave and it also needs a local rebase to fix whatever is conflicting with current dev HEAD but then it should be fine to go into dev.

File: framework/graphics/fps_info.cpp (hunk 10 above)
Comment (marius-pelegrin-arm) [reply framework/graphics/fps_info.cpp]: Done

File: framework/graphics/fps_info.cpp (hunk 12 above)
Comment (marius-pelegrin-arm) [reply framework/graphics/fps_info.cpp]: Done, but I just had to make kJsonIndentWidth visible for that (Put it as a static constexpr in the header)

File: framework/graphics/fps_info.cpp (hunk 11 above)
Comment (marius-pelegrin-arm) [reply framework/graphics/fps_info.cpp:134]: Yes, it was deliberate, this is not needed, was wrongly implemented, and would be a bit hard/annoying to implement here. Not worth it.

**Comment by andrew-lunarg** [comment]
Hi, this looks great, thank you. For future PRs, if you are kind enough to contribute some, we rebase feature branches on dev rather than merging dev into feature branches and creating merge commits.

I did the rebase and made the last changes myself over on a fresh branch / PR so I didn't erase history on your PR's branch by pushing the rebase to it. I preserved your authorship when doing so.

If this fresh PR passes CI it will go in today and I'll pop back here to close this one.
The PR: https://github.com/LunarG/gfxreconstruct/pull/1280

**Comment by andrew-lunarg** [comment]
Merged rebased version https://github.com/LunarG/gfxreconstruct/pull/1280

**Pull request closed without merging**